eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs

```diff
@@ -318,239 +655,266 @@ impl Eth2NearRelay {
+                &relay.eth_client_contract,
+            )
+            .unwrap();
+        assert_eq!(last_slot, blocks.1);
+        if let Ok(_) = relay.get_execution_block_by_slot(last_slot) {
+            panic!("Wrong last slot");
+        }
+    }
 
-        let finalized_slot = self.eth_client_contract.get_finalized_beacon_block_slot()?;
-        trace!(target: "relay", "Finalized slot on near={}", finalized_slot);
+    #[test]
+    #[ignore]
+    fn try_submit_update_with_not_enough_blocks() {
+        let mut relay = get_relay(true, true);
+        let mut slot = get_finalized_slot(&relay);
+        slot += 1;
 
-        let last_submitted_slot = self.eth_client_contract.get_last_submitted_slot();
-        trace!(target: "relay", "Last submitted slot={}", last_submitted_slot);
+        send_execution_blocks(&mut relay, slot, 1099391);
 
-        let slot = max(finalized_slot, last_submitted_slot);
-        trace!(target: "relay", "Init slot for search as {}", slot);
+        let finalized_slot = get_finalized_slot(&relay);
+        assert_eq!(finalized_slot, 1099360);
 
-        return if self.enable_binsearch {
-            self.binary_slot_search(slot, finalized_slot, last_eth_slot)
-        } else {
-            self.linear_slot_search(slot, finalized_slot, last_eth_slot)
-        };
+        relay.send_light_client_updates(1099360);
+        let finalized_slot= get_finalized_slot(&relay);
+
+        assert_eq!(finalized_slot, 1099360);
     }
 
-    fn linear_slot_search(
-        &self,
-        slot: u64,
-        finalized_slot: u64,
-        last_eth_slot: u64,
-    ) -> Result<u64, Box<dyn Error>> {
-        return if slot == finalized_slot || self.block_known_on_near(slot)? {
-            Ok(self.linear_search_forward(slot, last_eth_slot))
-        } else {
-            Ok(self.linear_search_backward(finalized_slot, slot))
-        };
+    #[test]
+    fn test_not_invalid_attested_slot() {
+        let mut relay = get_relay(true, true);
+        let finalized_slot = 1099360;
+        let possible_attested_slot = finalized_slot
+            + ONE_EPOCH_IN_SLOTS * 2
+            + ONE_EPOCH_IN_SLOTS * relay.light_client_updates_submission_frequency_in_epochs;
+        if let Ok(_) = relay.get_execution_block_by_slot(possible_attested_slot) {
+            panic!("possible attested slot has execution block");
+        }
+
+        let attested_slot = relay.get_attested_slot(finalized_slot).unwrap();
+        relay.get_execution_block_by_slot(attested_slot).unwrap();
     }
 
-    fn binary_slot_search(
-        &self,
-        slot: u64,
-        finalized_slot: u64,
-        last_eth_slot: u64,
-    ) -> Result<u64, Box<dyn Error>> {
-        return if slot == finalized_slot || self.block_known_on_near(slot)? {
-            self.binsearch_slot_forward(slot, last_eth_slot + 1)
-        } else {
-            self.binsearch_slot_range(finalized_slot, slot)
-        };
+    #[test]
+    #[should_panic]
```

Comment:
Ideally, to specify the reason for panicking

Reply:
fixed: https://github.com/aurora-is-near/rainbow-bridge/pull/787/commits/939a0aa482fff0062b127ae02d3025cbda6978c1